src/core/meshcop/border_agent.cpp

```diff
@@ -0,0 +1,600 @@
+
+exit:
+    (void) aMessageInfo;
+    if (error != kThreadError_None && message != NULL)
+    {
+        message->Free();
+    }
+    otLogFuncExit();
+}
+
+void BorderAgent::HandleRelayReceive(void *aContext, otCoapHeader *aHeader, otMessage *aMessage,
+                                     const otMessageInfo *aMessageInfo)
+{
+    static_cast<BorderAgent *>(aContext)->HandleRelayReceive(
+        *static_cast<Coap::Header *>(aHeader), *static_cast<Message *>(aMessage),
+        *static_cast<const Ip6::MessageInfo *>(aMessageInfo));
+}
+
+void BorderAgent::HandleRelayReceive(Coap::Header &aHeader, Message &aMessage,
+                                     const Ip6::MessageInfo &aMessageInfo)
+{
+    ThreadError error = kThreadError_None;
+    Coap::Header header;
+    Message *message = NULL;
+    Ip6::MessageInfo messageInfo;
+
+    otLogFuncEntry();
+    otLogInfoMeshCoP(GetInstance(), "received Relay Recevie from Joiner Router");
+
+    header.Init(kCoapTypeConfirmable, kCoapRequestPost);
+    header.SetToken(aHeader.GetToken(), aHeader.GetTokenLength());
+    header.AppendUriPathOptions(OPENTHREAD_URI_RELAY_RX);
+    header.SetPayloadMarker();
+    
+    VerifyOrExit((message = mNetif.GetSecureCoapServer().NewMeshCoPMessage(header)) != NULL, error = kThreadError_NoBufs);
+
+    SuccessOrExit(error = CopyPayload(aMessage, message));
+
+    messageInfo.SetPeerAddr(mCommissionerAddr);
+
+    otLogInfoMeshCoP(GetInstance(), "send Relay Receive to Commissioner");
+
+    SendCommissionerRelayReceive(*message, messageInfo);
+
+    
+exit:
+    (void)aMessageInfo;
+    if (error != kThreadError_None && message != NULL)
+    {
+        message->Free();
+    }
+    otLogFuncExit();
+}
+
+
+ThreadError BorderAgent::SendCommissionerRelayReceive(Message &aMessage, const Ip6::MessageInfo &aMessageInfo)
+{
+    ThreadError error = kThreadError_None;
+
+    otLogFuncEntry();
+
+    SuccessOrExit(error = mNetif.GetSecureCoapServer().SendMessage(aMessage, aMessageInfo));
+
+    otLogInfoMeshCoP(GetInstance(), "sent Realy Receive to Commissioner");
+
+exit:
+    (void) aMessageInfo;
+    otLogFuncExitErr(error);
+    return error;
+}
+
+void BorderAgent::HandleTimer(void *aContext)
+{
+    static_cast<BorderAgent *>(aContext)->HandleTimer();
+}
+
+void BorderAgent::HandleTimer(void)
+{
+    ThreadError error = kThreadError_Error;
+    PSKcTlv *pskc;
+    char aThreadPSKc[] = 
+    {
+        0xc3, 0xf5, 0x93, 0x68, 0x44, 0x5a, 0x1b, 0x61,
+        0x06, 0xbe, 0x42, 0x0a, 0x70, 0x6d, 0x4c, 0xc9,
+
+    };
+
+    otLogFuncEntry(); 
+
+    pskc = static_cast<PSKcTlv *>(mNetif.GetActiveDataset().GetNetwork().Get(Tlv::kPSKc));
+    if (pskc != NULL)
+    {
+        memcpy(aThreadPSKc, pskc->GetPSKc(), sizeof(aThreadPSKc));
+    }
+
+    SuccessOrExit(error = mNetif.GetSecureCoapServer().SetPsk(reinterpret_cast<const uint8_t *>(aThreadPSKc),
+                                                              static_cast<uint8_t>(sizeof(aThreadPSKc))));
```

Comment:
`sizeof(aThreadPSKc)` may be changed to a meaningful macro or const int value for it's fixed for Thread.